editorial: Modernize and replace prose with algorithms in BatteryManager

rakuco opened the pull request:
Similarly to #45, make the `BatteryManager` section more conformant to
modern spec writing practices, while avoiding making user-visible changes as
much as possible:
- Define a task source for this spec, and switch from "queue a task" to the
  more specific and recent "queue a global task" algorithm.
- Replace a lot of prose with internal slots and algorithms.
- Make the interface attributes simply return the values of the internal
  slots rather than defining logic in their getters.
- Define an actual algorithm for updating internal slots and dispatching
  events. The existing prose just said "the UA must change the attribute and
  fire an event" without really defining how to reach a `BatteryManager`
  object in the first place.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/rakuco/battery/pull/46.html" title="Last updated on Sep 8, 2021, 8:16 AM UTC (65d32d6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/battery/46/eb5b629...rakuco:65d32d6.html" title="Last updated on Sep 8, 2021, 8:16 AM UTC (65d32d6)">Diff</a>

Commit:
editorial: Modernize and replace prose with algorithms in BatteryManager

Similarly to #45, make the `BatteryManager` section more conformant to
modern spec writing practices, while avoiding making user-visible changes as
much as possible:
- Define a task source for this spec, and switch from "queue a task" to the
  more specific and recent "queue a global task" algorithm.
- Replace a lot of prose with internal slots and algorithms.
- Make the interface attributes simply return the values of the internal
  slots rather than defining logic in their getters.
- Define an actual algorithm for updating internal slots and dispatching
  events. The existing prose just said "the UA must change the attribute and
  fire an event" without really defining how to reach a `BatteryManager`
  object in the first place.

rakuco commented:
@reillyeon @marcoscaceres 
This one is bigger than #45, so some extra pairs of eyes are appreciated.

reillyeon commented on index.html:
```diff
@@ -240,8 +240,8 @@ <h3>
               <li>If [=this=].{{Navigator/[[BatteryManager]]}} is `null`, then
-              set it to the result of [=creating a new `BatteryManager`
-              object=] in [=this=].[=relevant realm=].
+              set it to the result of creating a new `BatteryManager` object in
```
```suggestion
              set it to the result of creating a [=new=] {{BatteryManager}} in
```

rakuco replied:
Done.

reillyeon commented on index.html:
```diff
@@ -285,168 +271,276 @@ <h2>
+      </aside>
       <section>
-        <h2>
-          Event handlers
```
Can you keep a version of this section around so that the event names are still references?

rakuco replied:
Done.

marcoscaceres commented:
Will try to review soon!

anssiko commented:
I authorize @reillyeon to merge this PR once we get @marcoscaceres's eyeballs on it.

marcoscaceres commented on index.html:
```diff
@@ -285,107 +271,211 @@ <h2>
-        </h2>
+      <aside class="note">
+        If the user agent is <a>unable to report the battery status
+        information</a>, the <a>BatteryManager</a>'s internal slots will remain
```
I'd suggest doing this globally: 

```suggestion
        information</a>, the {{BatteryManager}}'s internal slots will remain
```

It just helps distinguish between concepts and WebIDL things.

rakuco replied:
Done.

marcoscaceres commented on index.html:
```diff
@@ -285,107 +271,211 @@ <h2>
+          </tbody>
+        </table>
+        <section>
+          <h4>
```
```suggestion
          <h4 id="charging-internal-slot">
```

rakuco replied:
Is this still necessary after removing the descriptions column from the internal slots table?

marcoscaceres replied:
no, not necessary unless you want to reference it.

marcoscaceres commented on index.html:
```diff
@@ -285,107 +271,211 @@ <h2>
+                `true`
+              </td>
+              <td>
+                Charging state of the system's battery.
```
maybe this should like to the section below? 

```suggestion
                Charging state of the system's battery. See [[[#charging-internal-slot]]].
```

Same with others... alternatively. Why not move the sections below into here? Or vice versa. Seems a bit odd to have them almost defined twice.

rakuco replied:
I've removed the non-normative description column from the table, so that the descriptions are all gathered in the section below.

marcoscaceres commented on index.html:
```diff
@@ -285,107 +271,211 @@ <h2>
+          </h4>
+          <p>
+            The {{BatteryManager/[[Level]]}} internal slot represents the level
+            of the system's battery. It MUST be set to 0 if the system's
+            battery is depleted and the system is about to be suspended, and to
+            1.0 if the battery is full, the implementation is unable to report
```
Don't need to fix this now, but this is a bit confusing... it doesn't say when it must be set to 0 or 1? I guess when it's initialized... but this should really be part of some initialization steps.

marcoscaceres replied:
Or, when initialized, run "update the battery status and notify" and set them there.

rakuco replied:
My understanding is that this is automatically set to 1 when an instance is created (per the initial value specified in the internal slots table). Then once the UA decides to check the battery level, it will run 'update the battery status and notify' with the new level, be it 0 or something else.

Did you have something like always running the 'update status and notify' algorithm on object construction? If so I think it is better to change this separately since it sounds like a user-visible change (and if we decide to make user-visible changes we might as well do something more useful with the promise we return in `getBattery()`, whose algorithm is essentially synchronous at the moment).

marcoscaceres replied:
yeah, let's do that as followup... fwiw, I'm imagining this prose need to go into some initializer steps... as an example:

https://w3c.github.io/geolocation-api/#constructing-a-geolocationposition

rakuco replied:
Thanks, I've filed #48 to track this.

marcoscaceres commented on index.html:
```diff
@@ -411,43 +501,103 @@ <h2>
+          </li>
+        </ol>
+      </section>
+      <section id="multiple-batteries">
```
As a followup, I wonder if this should be actually integrated into the definition of each slot. That way, it's all kept together.

Single battery, do x... 

Multiple batteries, do y...

rakuco replied:
Good point. I've filed #47 to track this.

marcoscaceres reviewed (approved):
Overall looks good @rakuco! Just some suggestions, but a vast improvement! keep it up!

Commit:
Use [= =] notation in the update battery status and notify algorithm.

Commit:
Remove non-normative description of BatteryManager's internal slots.

The contents are essentially the same as the normative descriptions in the
subsequent section.

Commit:
Use {{BatteryManager}} to refer to it in the text.

Refer to the Web IDL interface rather than the BatteryManager concept.

reillyeon commented on index.html:
```diff
@@ -285,107 +271,194 @@ <h2>
+            must run the <a>update the battery status and notify</a> algorithm
+            with {{BatteryManager/[[Charging]]}}, `true` or `false` depending
+            on whether the battery is charging or discharging, and
+            "`chargingchange`".
```
Link to the definition of `"chargingchange"` here, and the same for the other events below.

rakuco replied:
Done. I was following the WHATWG style from e.g. https://notifications.spec.whatwg.org/#object-members that does not define the items in the second column of the event handler table. I've now switched to `<dfn>`ing both columns, PTAL.

Commit:
BatteryManager: Define and reference the event types in the Event Handler table.

rakuco commented:
thanks for reviewing, @reillyeon and @marcoscaceres 

I'd appreciate it if someone could merge this PR

anssiko commented:
Thanks @rakuco! You should add yourself to the editors if you plan to keep an eye on this spec. Otherwise, we’ll give you huge credits in the ack section for these modernizations.

rakuco commented:
So far I've just been trying to catch up with the backlog of things I'd promised to do back in April during the DAS WG virtual meeting :-) Once I finish making the spec look better (and hopefully address #15) I can consider officially becoming an editor.
